software/bsg_manycore_lib/bsg_tile_group_barrier.hpp

```diff
@@ -0,0 +1,260 @@
+//====================================================================
+// bsg_tile_group_barrier.hpp
+// 02/14/2019, shawnless.xie@gmail.com
+// 02/19/2020, behsani@cs.washington.edu
+//====================================================================
+// The barrier implementation for tile group in manycore
+// Usage:
+//      1. #include "bsg_tile_group_barrier.hpp"
+//      1. bsg_barrier<Y dimension, X dimension> my_barrier;
+//      3. my_barrier.sync(); 
+//
+//Memory Overhead
+//       (2 + X_DIM + 4) + (2 + Y_DIM +4) 
+//      =(12 + X_DIM + Y_DIM) BYTES
+//
+//Worst Case Performance:
+//      1. row sync     :   X_DIM     cycles //all tiles writes to center tile of the row
+//      2. row polling  : 3*X_DIM     cycles // lbu <xx>; beqz <xxx;
+//      3. col sync     :   Y_DIM     cycles //all tiles writes to the center tiles of the col
+//      4. col polling  : 3*Y_DIM     cycles // lbu <xx>; beqz <xx>;
+//      5. col alert        Y_DIM     cycles // store
+//      6. row alert        X_DIM     cycles // store
+//      -----------------------------------------------
+//                        5*( X_DIM + Y_DIM)
+//      For 3x3 group,  cycles = 181, heavy looping/checking overhead for small groups.
```

Comment:
This is a great comment!!

Reply:
Yes it is! But it's not mine I copied it from Shaolin's code :D